blog/2020/12/package_latency.md

```diff
@@ -0,0 +1,718 @@
+
+This post is based on DataFrames 0.22.1, and version 0.9 of the underlying CategoricalArrays.  If you follow the steps of this blog post with different versions, you're likely to get different results from those shown here, partly because many of the issues we identified have been fixed in more recent releases.  It should also be emphasize that these analysis tools are only supported on Julia 1.6 and above; at the time of this post, Julia 1.6 not yet to "alpha" release phase but can be obtained from [nightly] snapshots or built from [source].
+
+## Taking a baseline measurement
+
+A good first step is to roughly document the current compiler latency costs.  For a package like DataFrames that contains an extensive test suite, an easy way to do this is to time the entire test suite:
+
+```julia
+shell> pwd
+/home/tim/.julia/dev/DataFrames/test
+
+julia> using DataFrames; @time include("runtests.jl")
+Running tests:
+⋮                     # lots of suppressed lines of output
+620.652994 seconds (1.25 G allocations: 72.799 GiB, 3.35% gc time, 96.29% compilation time)
+```
+
+Julia told us that a whopping 96.29% of the time was spent on compilation. It's pretty common among Julia packages that compilation is the dominant cost of running the test suite. Unsurprisingly, it's considerably faster if we just run the suite a second time within the same session:
+
+```julia
+julia> @time include("runtests.jl")
+⋮
+235.602932 seconds (396.67 M allocations: 23.434 GiB, 3.05% gc time, 89.01% compilation time)
+```
+
+While this is much faster, you'll note that compilation still accounts for almost 90% of the time. Why is there any recompilation at all?  One reason is that tests that (re)define functions, including lines like
+
+```julia
+map(x->x^2, a)   # when run from "top level," this defines a new anonymous function
+```
+force a new round of specialization of the `map` method.  DataFrames has a *lot* of tests that look like this.
+
+Having obtained this simple baseline measurement, let's get started.
+
+## Identifying the most costly-to-infer methods
+
+Our first goal is to identify methods that cost the most in inference.
+To do this, start from a fresh Julia session and do the following:
+
+```julia
+shell> pwd
+/home/tim/.julia/dev/DataFrames/test
+
+julia> using SnoopCompile
+
+julia> tinf = @snoopi_deep include("runtests.jl");
+Running tests:
+⋮
+```
+
+`@snoopi_deep` is a new tool in [SnoopCompile] which leverages new functionality in Julia.  Like the older `@snoopi`, it measures what is being inferred and how much time it takes.  However, `@snoopi` measures aggregate time for each "entrance" into inference, and it includes the time spent inferring all the methods that get inferrably dispatched from the entrance point.  In contrast, `@snoopi_deep` extracts this data for each method instance, regardless of whether it is an "entrance point" or called by something else.
```

Comment:
Perhaps "extracts this data for each method instance" -> "extracts the time spent inferring each method instance _exclusive_ of the time spent inferring other method instances" or similar? :)